### PR TITLE
several small improvements to Regex

### PR DESCRIPTION
**Pull request by bjmllr**
* sort Regex methods
* convert a trivial getter (`options`) to use the getter macro
* document all public Regex methods
* add class overview documentation for Regex
* separate Regex specs from MatchData specs
* sort Regex and MatchData specs
* change the meaning of `Regex#to_s` to match that of Ruby
* add `Regex.union` and an equivalent operator, `Regex#+`

My original intention was only to add `Regex.union`, but I have resolved not to touch a class without documenting all its methods first!

Ruby pretty much completely documents Onigmo regular expression syntax with `Regexp`, but it seemed unwise to try to create such complete documentation for the regex syntax, when we can link to the existing (and I think quite good) PCRE docs. A weakness of PCRE is that it seems to lack a good introduction to regular expressions, presumably because Perl has an excellent one, so I wrote the class overview with the intention of filling that small gap for Crystal.

The more Ruby-ish behavior for `Regex#to_s` doesn't appear to break anything in the compiler, and it allows for one Regex to be transparently interpolated in another while retaining its flags. It does expose a small annoyance, which is that the PCRE options letters (imsx) are not exactly the same as Crystals's options letters (imx) - since we use groups inside the Regex like `(?i-msx:foo)` to preserve options, and these are ultimately passed to PCRE unchanged, we have to use the PCRE options letters. At this point, would it be better to use PCRE's options letters everywhere? Crystal combines PCRE_MULTILINE (m) with PCRE_DOTALL (s) to match Ruby behavior, but Crystal no longer has a goal of being source compatible with Ruby, and it might be beneficial to someone to be able to use PCRE_MULTILINE independently of PCRE_DOTALL or vice versa (I've never wanted it).

Because `Regex.union` (and `Regex#+`) is based on interpolation, it's possible that it will try to create an invalid regex and cause an ArgumentError to be raised at runtime. Since regex literals are syntax-checked at compile time and string components are escaped, I think the only time this can happen is when there are multiple same-named capture groups across two or more of the regexes to be combined, so I've tried to make this point clear in the docs.

### CODE REVIEW

File: src/regex/regex.cr [hunk 1]
```diff
@@ -110,43 +396,107 @@ class Regex
...
+  # Convert to String in subpattern format. Produces a String which can be
+  # embedded in another Regex via interpolation, where it will be interpreted
+  # as a non-capturing subexpression in another regular expression.
+  #
+  # ```
+  # re = /A*/i                 #=> /A*/i
+  # re.to_s                    #=> "(?i:A*)"
+  # "Crystal".match(/t#{re}l/) #=> #<MatchData "tal">
+  # re = /A*/                  #=> "(?:A*)"
+  # "Crystal".match(/t#{re}l/) #=> nil
```
Comment (jhass) [inline src/regex/regex.cr]: I think the examples miss the always inserted `-`.

File: src/regex/regex.cr [hunk 2]
```diff
@@ -1,5 +1,182 @@
...
+# /asdf\nz/im =~ "ASDF\nZ" #=> 0
+# ```
+#
+# PCRE supports other encodings, but Crystal strings are UTF-8 only, so Crystal
+# regular expressions are also UTF-8 only (by default).
+#
+# PCRE optionally permits named capture groups (named subpatterns) to not be
+# unique. Crystal exposes the name table of a Regex as a
+# Hash of String => Int32, and therefore requires named capture groups to have
+# unique names with a single Regex.
```
Comment (jhass) [inline src/regex/regex.cr]: "within"?

File: src/regex/regex.cr [hunk 3]
```diff
@@ -34,74 +231,163 @@ class Regex
...
-      name_offset = capture_offset + 2
-      name = String.new( (name_table + name_offset).pointer(name_entry_size-3) )
-
-      lookup[capture_number] = name
+  # Determines Regex's source validity. If it is, `nil` is returned.
+  # If it's not, a String containing the error message is returned.
+  #
+  # ```
+  # Regex.error("(foo|bar)") #=> nil
+  # Regex.error("(foo|bar") #=> "missing ) at 8"
```
Comment (jhass) [inline src/regex/regex.cr]: Missing `?` in the examples.

File: src/regex/regex.cr [hunk 4]
```diff
@@ -34,74 +231,163 @@ class Regex
...
+  # indexes offset.
+  #
+  # ```
+  # re = Regex.union([/skiing/i, "sledding"])
+  # re.match("Skiing")   #=> #<MatchData "Skiing">
+  # re.match("sledding") #=> #<MatchData "sledding">
+  # ```
+  def self.union(patterns : Array(Regex | String))
+    new patterns.map { |pattern| union_part pattern }.join("|")
   end
```
Comment (jhass) [inline src/regex/regex.cr]: I'd say accept an `Enumerable(Regex|String)` and use a tuple in the example.

File: src/regex/regex.cr [hunk 5]
```diff
@@ -34,74 +231,163 @@ class Regex
...
+  # indexes offset.
+  #
+  # ```
+  # re = Regex.union(/skiing/i, "sledding")
+  # re.match("Skiing")   #=> #<MatchData "Skiing">
+  # re.match("sledding") #=> #<MatchData "sledding">
+  # ```
+  def self.union(*patterns : Regex | String)
+    new patterns.map { |pattern| union_part pattern }.join("|")    
+  end
```
Comment (jhass) [inline src/regex/regex.cr]: With the above change you should be able to forward to it just fine here.

**Comment by bjmllr** [comment]
@jhass thanks for the feedback, I've incorporated all of your suggestions.

**Comment by asterite** [comment]
@bjmllr This... is... AMAZING!!

I don't think this class could have been documented in a better way. Also thank you for the `union`, `+` and other methods. I have to confess that doing this in Ruby:

```
 irb(main):001:0> a = /x/
=> /x/
irb(main):002:0> r = /#{a}/
=> /(?-mix:x)/
```

I had no idea what that "?-mix:" meant... :blush: Now I know :smile_cat: 

We use PCRE because it was the simplest way to have Regex support in the language. But we can always decide to change the internal implementation, for example to use Onigmo, if we find it better. So I wouldn't tie the docs to PCRE, but for now it's OK (I don't know when, or if, we are going to change PCRE).

That's why, regardless of how PCRE behaves, we should have our set of Regex options and translate those to the internal implementation.

**Comment by jhass** [comment]
I think clearly documenting the regex engine is a good choice. Most have at least some details implemented differently or at least different feature sets, so it's not about the options, swapping out the engine will always be a breaking change for *something*, you can't abstract that.

**Comment by jhass** [comment]
That said I wouldn't mind using Oniguruma/Onigmo instead, simply because I know its behaviour best :P Afaik that's also where MRI gets its encoding support from, so that might be something to gain at the same time.

**Comment by asterite** [comment]
@bjmllr I made a small change in the docs for `options` (see above commit). It returns an `Enum`, but the string representation is what you documented ("IGNORE_CASE, EXTENDED"), basically [this](https://github.com/manastech/crystal/blob/master/src/enum.cr#L4-L40) (that code is a bit convulted, but [works](https://github.com/manastech/crystal/blob/master/src/enum.cr#L4-L40)).

@jhass Thanks for reviewing pull requests!

**Comment by bjmllr** [comment]
Right, the string representation is misleading. I wasn't sure how to show an Enum result in an example, now I know. :-)

**Comment by asterite** [comment]
I actually like that you can bind to C libraries, define enums, and when you debug the code you get a nice, understandable string representation, rather than a numeric code that I have to map back to meaning :-)

By the way, we borrowed that feature from C# :-)

**Comment by bjmllr** [comment]
Sorry, I meant that the way I put quotes around it was misleading, to someone reading it. Also doesn't help that I put the wrong type and method name there, bleh.

I didn't mean that the compiler was misleading, the human readable Enum representation is awesome. :-)

**Pull request merged**